### PR TITLE
lib: lte_link_control: Add support for %XT3412 AT notifications

### PR DESCRIPTION
**Pull request by simensrostad**
Add support for %XT3412 AT notifications that allows the application
to get a pre-warning before a TAU is performed by the modem.

Closes CIA-102

**This PR also adds an API to ```at_cmd_parser``` that retrieves signed 64 bit values from AT responses. %XT3412 notifications can contain a <time> parameter exceeding whats possible to retrieve with the current API.

### CODE REVIEW

File: include/modem/at_params.h [hunk 1]
```diff
@@ -38,6 +38,8 @@ enum at_param_type {
 	AT_PARAM_TYPE_INVALID,
 	/** Parameter of type short. */
 	AT_PARAM_TYPE_NUM_SHORT,
+	/** Parameter of type long. */
+	AT_PARAM_TYPE_NUM_LONG,
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply include/modem/at_params.h:42]: Agreed. Lets be exact here. Ill update the types/functions to use ```int64``` specifier.

**Comment by rlubos** [comment]
There is a parallel PR reworking the `at_params` module in terms of the unsigned short/int handling, I think we should keep this PR in sync with https://github.com/nrfconnect/sdk-nrf/pull/4080.
In general, in #4080:
* The actual numeric at_param type is limited to `AT_PARAM_TYPE_NUM_INT` given all int-type params end up in the same bucket anyway (int32_t for now, int64_t after changes here),
* The actual value stored in the integer is verified in the getter function - if it does not fit in the storage provided in the getter, an error is returned. 

Given the broader audience in this PR, please feel free to review the other PR and post your opinion if you disagree with the approach. As I mentioned before, we should keep the changes in sync in order to avoid design conflicts. If there're no objections for #4080 to get in I think we should merge it first given it was submitted earlier (it already has two approvals).

**Comment by simensrostad** [comment]
> There is a parallel PR reworking the `at_params` module in terms of the unsigned short/int handling, I think we should keep this PR in sync with #4080.
> In general, in #4080:
> 
> * The actual numeric at_param type is limited to `AT_PARAM_TYPE_NUM_INT` given all int-type params end up in the same bucket anyway (int32_t for now, int64_t after changes here),
> * The actual value stored in the integer is verified in the getter function - if it does not fit in the storage provided in the getter, an error is returned.
> 
> Given the broader audience in this PR, please feel free to review the other PR and post your opinion if you disagree with the approach. As I mentioned before, we should keep the changes in sync in order to avoid design conflicts. If there're no objections for #4080 to get in I think we should merge it first given it was submitted earlier (it already has two approvals).

I don't see any immediate issues with implementing the changes in this PR on top of https://github.com/nrfconnect/sdk-nrf/pull/4080.

**Comment by rlubos** [comment]
Thank you @simensrostad, I'll take the liberty to merge #4080 given it was open for some time already and is good to merge.

**Comment by simensrostad** [comment]
Rebased and aligned according to new changes in lib ```at_params```.

File: lib/at_cmd_parser/Kconfig [hunk 2]
```diff
@@ -6,4 +6,4 @@
 
 config AT_CMD_PARSER
 	bool "AT command parser library"
-
+	depends on NEWLIB_LIBC || EXTERNAL_LIBC
```
Comment (rlubos) [inline lib/at_cmd_parser/Kconfig:9]: Why do we need to depend on NEWLIB? Is it `strtoll()`?

**Review by rlubos — APPROVED** [review]
`at_params` changes look fine, on the XT3412 I cannot comment on the details but it seems reasonable. I'm not in favour of adding NEWLIB dependency, but I don't see an other way around for now since we need to parse 64-bit integer. We could probably consider adding `strtoll` to the minimal libc in the future.

**Review by jtguggedal — COMMENTED** [review]
Looks good. Left a few comments on documentation

File: include/modem/lte_lc.h [hunk 3]
```diff
@@ -156,6 +156,9 @@ enum lte_lc_evt_type {
 	 *  LTE mode is currently used by the modem.
 	 */
 	LTE_LC_EVT_LTE_MODE_UPDATE,
+
+	/** Tracking Area Update pre-warning. */
```
[the remark this answers is not in the document]
Comment (simensrostad) [reply include/modem/lte_lc.h:160]: Good idea. Ill update the docs.

**Comment by simensrostad** [comment]
@jtguggedal Updated the docs.

**Review by jtguggedal — APPROVED** [review]
Looks good to me, thanks

File: lib/at_cmd_parser/Kconfig (hunk 2 above)
Comment (lemrey) [reply lib/at_cmd_parser/Kconfig:9]: We should have `#ifdef`-ed the new functionality of this library that requires `newlibc`, and enabled that were necessary per-sample. As it is done right now, it is forcing a `libc` other than `minimal-libc` on all users of the parser, which is not ideal.

**Pull request merged**